### DISCUSSION TITLE
Can we get du automatically without explicitly defining it as a variable

### DISCUSSION THREAD

**Discussion by cylin928** [do-mpc/do-mpc#356] [Q&A]
I have a discrete system.
The rterm is calculated by du'Rdu according to the document.
https://www.do-mpc.com/en/latest/api/do_mpc.controller.MPC.set_rterm.html

I was wondering can we call du directly without defining it to set up a constraint, e.g., du <= 0.1?

Or, we need to modify the system equation explicitly to call du (in this case we define a variable called du).

Thank you very much!

**Comment by 4flixt** [comment]
Hi @philip928lin ,

``du`` is not introduced as a symbolic variable and can therefore not easily be obtained. As you suggested, it can be introduced to the model. Alternatively, you can call [mpc.prepare_nlp](https://www.do-mpc.com/en/latest/api/do_mpc.controller.MPC.prepare_nlp.html#do_mpc.controller.MPC.prepare_nlp) after which all MPC optimization variables can be accessed with [mpc.opt_x](https://www.do-mpc.com/en/latest/api/do_mpc.controller.MPC.opt_x.html).
This is a bit advanced but allows to form ``du`` youself.

Best
4flixt

**Comment by cylin928** [comment]
Thank you @4flixt !
I am able to solve my issue by adding the following code.
`
# Add customized constraints 
mpc.prepare_nlp()
ulist = mpc.opt_x['_u']
du = 0.1

for i in range(len(ulist)):
    # Create new constraint
    if i == 0:
        extra_cons = mpc.opt_x['_u', i, 0] - mpc.opt_p['_tvp', i, 'U0t']
    else:
        extra_cons = mpc.opt_x['_u', i, 0] - mpc.opt_x['_u', i-1, 0]
    mpc.nlp_cons.append(extra_cons)
    # Create appropriate upper and lower bound 
    shape = extra_cons.shape
    mpc.nlp_cons_lb.append(-np.array([du]*shape[0]).reshape(shape))
    mpc.nlp_cons_ub.append(np.array([du]*shape[0]).reshape(shape))

mpc.create_nlp()
`

**Reply by bdcoyle** [reply]
Hi @philip928lin , sorry to dig up this old discussion. We are trying to implement the same thing as you, but are having problems with the tvp.
We try to pass our previous u value via the tvp_fun, but it doesn't seem to work. No matter what value we pass in via the tvp_fun, it seems the solver always treats it as though it is '0' in constraint expression (which is formulated the same as yours with 'extra_cons').
Even if we just pass a fixed scalar value (instead of previous power), it seems it is always treated as '0' by the solver.

Did you notice anything like this in your development? If possible, could you maybe share your tvp_fun as well?
Many thanks!

**Comment by 4flixt** [comment]
nice, well done :)
